README.md
# pcgs_scraper: Tools for scraping coin data from PCGS

![pcgs_scraper_logo](pcgs_logo_and_scraper.png)

[![CodeFactor Grade](https://img.shields.io/codefactor/grade/github/ryanamannion/pcgs_scraper?color=brightgreen)](https://www.codefactor.io/repository/github/ryanamannion/pcgs_scraper/overview/main)
[![GitHub release](https://img.shields.io/github/v/release/ryanamannion/pcgs_scraper?include_prereleases)](https://github.com/ryanamannion/pcgs_scraper/releases)
[![GitHub code size in bytes](https://img.shields.io/github/languages/code-size/ryanamannion/pcgs_scraper)](https://github.com/ryanamannion/pcgs_scraper)
[![license](https://img.shields.io/badge/license-CC0-brightgreen)](https://github.com/ryanamannion/pcgs_scraper/blob/main/LICENSE.txt)
[![GitHub last commit (branch)](https://img.shields.io/github/last-commit/ryanamannion/pcgs_scraper/main)](https://github.com/ryanamannion/pcgs_scraper/commits/main)

Scrape current PCGS coin prices from https://www.pcgs.com/prices and save them to a lookup table for easy price lookup 
or other manipulation

This repo is not sponsored or endorsed by PCGS. Logo for stylistic purposes, following PCGS Brand Guidelines

If you use this repo, send me an email! I'd love to hear how you used it and what I can improve

## Quick Start

### Install locally to scrape and query prices
1. Clone the repository to the directory of your choice with `$ git clone https://github.com/ryanamannion/pcgs_scraper.git`
2. `$ cd pcgs_scraper`
3. If you are using a venv or other environment, activate it
4. `$ pip install .`
5. Navigate to the pcgs_scraper subdirectory
6. `$ python scraper.py`
7. `$ python pcgs_query.py -q '1909-S VDB Cent'`

### Install with pip as a package
1. Download the latest release from `Releases` on GitHub
2. Activate your environment
3. `$ pip install pcgs_scraper-X.Y.Z.tar.gz`
4. Use pcgs_scraper functions in your own scripts
5. If you need to access the downloaded files:
    * find directory with `$ pip list -v | grep pcgs`
    * copy directory path and `cd` to it

## Design and Functionality

### `scraper.py`

`scraper.py` is the main file in this library, and handles the dispatching of the scraping scripts. Additionally, this file handles postprocessing of those scraping scripts.

Running it from the command line calls the cli() function.
The interface prompts the user to download the necessary files if they have not been scraped already. It will download
both the PCGS#-->Description information from www.pcgs.com/pcgsnolookup as well as the PCGS#-->Price information from
www.pcgs.com/prices. 

`scraper.py` then postprocesses that data and combines them to create a data-rich free table
(list of dictionaries) where each list item represents a coin. The free table has details about each coin, including the 
PCGS Number, Year, Denomination, Mint Mark (if applicable), Detail information (e.g. Full Bands or other details 
relevant to that particular coin), Price data at time of scraping, and metadata for the purposes of debugging (e.g. the URL it was
scraped from, etc.). 

Please note, during this step entries from the price data which do not have a description from the number data are 
excluded. These are mostly type coins, as well as type sets and other subsets of coins which can be given a valuation.

The final free table is saved to `data/pcgs_price_guide.{pkl, json}`, based on what the user selects in the CLI

### `pcgs_prices.py`

The first step in creating the price guide is to scrape the prices from www.pcgs.com/prices. 

The PCGS coin prices website is a labyrinth of html. This script first navigates to https://www.pcgs.com/prices, where
it navigates through each category (e.g. Type Coins, Half-Cents and Cents, etc.) and saves urls for each subcategory. 
The script then follows each subcategory URL and scrapes all price information from the table, including the PCGS# as
well as the prices for each grade. The website divides the grades into different pages, or bins of grades: 1-20, 25-60,
and 61-70. That means that for each subcategory there are three pages to scrape data from. (Note: I skip the 
"Most Active" page because it is redundant).

In order to ensure that a rogue error at a later step won't cause the user to lose all the data from scraping, which can
take some time, the data is saved to a pickle file at the end of the preliminary scraping function, and before the 
processing step that combines the data from the three bins into one lookup table. 
This file is saved in the pcgs_scraper directory using the date and time upon
completion to name the file `data/pcgs_prices-DD-MM-YYY-HH:MM:SS.pkl`. This file serves as the input to the 
processing function, which merges rows with the same PCGS# to creates the lookup table. It can be used any time with the 
`-p` command line option to be reprocessed should you need historical price data

The processing function saves the price data both as a pickle file and as a json file (because why not). These  files are saved to the same directory and named `data/scraped_pcgs_prices.{json, pkl}`

The resulting data structure is a dictionary. The keys are the PCGS Numbers, and the values are 
data extracted from the tables. 
- pcgs_num: (str) PCGS Number
- desig: (str) Designation (More info: https://www.pcgs.com/news/a-look-at-pcgs-designations)
- prices: (dict) Price by grade, each grade points to a tuple (n=2) of prices (each a str). The top price in the table 
  is index 0, the bottom is index 1. See Known Issues #1
- merged_from: the three entries for each grade bin used to create this merged entry (see paragraph one of this 
  section), retained for debugging purposes

```python
merged_entry = {
            'pcgs_num': pcgs_num,       # PCGS Number
            'desig': merged_desig,      # Designation (BN, RB, RD)
            'prices': price_by_grade,   # Price dictionary {Grade: [Price, Price+]}
            'merged_from': entries,     # History for merge, for debugging
        }
price_guide[pcgs_num] = merged_entry
```

### `pcgs_nums.py`

The second step in creating the price guide is to scrape the mappings of PCGS Numbers to detailed descriptions, which
adds high quality information about a coin's year, mint mark, denomination, and other details. This script uses the same
function from `pcgs_prices.py` to scrape URLs from the main page of the PCGS# lookup page by category and subcategory.
Each subcategory URL is then followed and the number:description pair is scraped and stored in a free table and saved to
`number_data.pkl`.


### `pcgs_query.py`

One a user has compiled the `pcgs_price_guide.pkl` binary, it can be queried from the command line with:

`$ python pcgs_query.py -q '1909-S VDB Wheat Cent'` 

or similar inputs. The query function uses some regex to determine the necessary elements of the query: the year and
denomination. It can then also find the mint mark (specified with -M where M stands for mint) to help narrow down
the search. Once the search algorithm has a target year and denomination (and possibly mint), it will rank the
results by Levenshtein Edit Distance from the user-generated input string. The user can then simply choose from the 
results list (if there are more than one option) and the price data will be displayed.

**Here are some general query guidelines:**

1. Always specify a year
2. Always specify a denomination
    * Can be of the form 1C, 3CS, 3 cent silver, $1, Dollar, half dollar, $2.50 etc.
3. If you want to specify a mint mark, do so with a hyphen following the year, e.g. `-q '1909-S VDB Cent'`


## Detailed Usage Notes:

### Running `pcgs_prices.py`
1. To show help dialogue: `$ python pcgs_prices.py --help`
2. To scrape all prices and clean up the data: `$ python pcgs_prices.py --all`
3. To just scrape data, create new unprocessed data binary: `$ python pcgs_prices.py --scrape_only`
    * This will save a file called `pcgs_prices-DD-MM-YYY-HH:MM:SS.pkl` with the current date and time
4. To just turn unprocessed binary into a lookup table: `$ python pcgs_prices.py --process path/to/pcgs_prices-DD-MM-YYY-HH:MM:SS.pkl`
    * This saves two files: `pcgs_price_guide.{json, pkl}`, both are of the same object 
    
### Running `pcgs_nums.py`

`pcgs_nums.py` has no CLI options. Running `$ python pcgs_nums.py` will download the number data and save it to 
`number_data.pkl`

### Running `pcgs_query.py`

1. Specify a query with `-q`
2. Specify a source price_guide binary with `-p`

## Known Issues and Future Changes:

**You can find current issues and enhancement ideas in the [`Issues`](https://github.com/ryanamannion/pcgs_scraper/issues) tab of GitHub**


CHANGELOG
# version 0.0.4
* adds verbosity option for query for use in MakeCents
* query exact match to description triggers return of that coin
* updates to README docs
* adds support for coinfacts narratives and images in number scraper
* fixes bug that made response None after a http 429 wait period

# version 0.0.3
* fixes typo in pcgs_nums that prevented nums from scraping

# version 0.0.2
* fixes import error stemming from the git repo, package, and file name being pcgs_scraper

# version 0.0.1
* first commit


pcgs_scraper/utils.py
#!/usr/bin/env python3
"""
utils.py

utils for scraping

Author: Ryan A. Mannion, 2020
github: ryanamannion
twitter: @ryanamannion
"""
import re
import sys
import time
import requests

from bs4.element import NavigableString


##################
# SCRAPING UTILS #
##################

def request_page(page_url):
    """
    Makes sure page is responding

    :param page_url: url to request
    :return: requested page if its working, error message with status if not
    """
    response = requests.get(page_url)
    if response.status_code == 429:
        # too many requests
        retry_after = response.headers['retry-after']
        print("Encountered response status 429: too many requests\n"
              f"Waiting {retry_after}s and retrying...")
        time.sleep(int(retry_after) + 2)     # for good measure
        print("Retrying...")
        response = request_page(page_url)
    elif not response.status_code == 200:
        print(f'Something went wrong with {page_url}!\n'
              f'Status code: {response.status_code}\n'
              f'Status text:\n' + response.text)
        sys.exit()
    return response


def non_ns_children(tag, search_type):
    """
    Filters out NavigableString children from tree navigation, allows use of
    .children and .descendants methods

    :param tag: bs4.Tag object
    :param search_type: string, {children, descendants}
    :return: list of bs4.Tag objects
    """
    filtered = []
    if search_type == 'children':
        for child in tag.children:
            if not isinstance(child, NavigableString):
                filtered.append(child)
    elif search_type == 'descendants':
        for child in tag.descendants:
            if not isinstance(child, NavigableString):
                filtered.append(child)
    return filtered


#########################
# QUERY & PARSING UTILS #
#########################
YEAR = re.compile(r'\(?(\d\d\d\d)\)?(/\(?\d+\)?)*')

MINT_CODES = [
    'P',        # Philadelphia, also often no mint mark
    'D',        # Dahlonega 1838-1861 (gold); Denver 1906-
    'S',        # San Francisco 1854-
    'CC',       # Carson City 1870-1893
    'O',        # New Orleans 1838-1909
    'W',        # West Point 1937- (commems)
    'M',        # Manila, Philippines 1920-1922, 1925-1941  (centavo)
    'A',
    'H'
    ]
mint_option = r'(' + r'|'.join(MINT_CODES) + r')'
#               -D                     /D
mint_pattern = r'-(' + mint_option + '(/' + mint_option + ')?)'
mint_pattern = r'-((P|D|S|CC|O|W|M|A|H)(/(P|D|S|CC|O|W|M|A|H))?)'
MINT = re.compile(mint_pattern)
MINT_CI = re.compile(mint_pattern, re.IGNORECASE)

DENOMINATIONS = [
        # US Coins
        r'1/2C',     # half cent
        r'1C',       # 1 cent
        r'Cent',
        r' C ',
        r'2C',       # 2 cent
        r'3CS',      # 3 cent silver
        r'3CN',      # 3 cent nickel
        r'5C',       # 5 cent (nickel)
        r'H10C',     # half Dime
        r'10C',      # dime
        r'20C',      # 20 cent
        r'25C',      # quarter
        r'50C',      # half dollar
        r'\$1(?!0)',       # dollar coin
        r'\$2.50',    # $2.50 gold
        r'\$3',       # $3 gold
        r'\$4',       # $4 gold
        r'\$5(?!0)',       # $4 gold
        r'\$10',      # $10 gold
        r'\$20',      # $20 St. Gaudens Double Eagle
        r'\$25',      # Gold Eagle
        r'\$50',      # Gold Eagle
        # Colonials and special cases
        r'\'?Penny\'?',
        r'1P',
        r'1/2 ?P',
        r'3Pence',
        r'2Pence',
        r'6Pence',
        r'Shilling',
        r'Shilng',    # Special case, see PCGS#249
        r'1/24RL',    # see PCGS#49
        r'9 Den',     # French colonies, Deniers
        r'15 Den',
        r'30 Den',
        r'Farth',     # e.g. PCGS#256
        r'Sou',       # French Colonies
        r'Sol',       # French Colonies, see PCGS#167113
        r'1/2 Db',    # see PCGS#489
        r'1/2 R'      # see PCGS#600506
        r'1/2 RL',
        r'Rial'
    ]
denom_option = r'(' + r'|'.join(DENOMINATIONS) + r')'
DENOM = re.compile(denom_option)
DENOM_CI = re.compile(denom_option, re.IGNORECASE)


def parse_descriptions(price_guide):
    """
    Parse descriptions for coins, e.g.
        '1794 1/2C Low Relief Head, BN' -->
            {'year': '1794',
             'denom': '1/2c',
             'desc': Low Relief Head,
             'color': BN
            }

    :param price_guide: price guide with descriptions, i.e.
        output of combine_number_price
    :return price_guide_parsed: original price guide but with new keys from the
        description
    """
    #################
    # PARSING NOTES #
    #################

    # year:
    #   #93924: 1914/(3) 5C                                 overstamp with paren
    #   #3939: 1918/7-D 5C                                   overstamp w/o paren
    #   #38975: 1825/4/(2) 25C Browning 2                       triple overstamp
    #   #715594: (2019) 1C Blank Planchet Explore & Discover Set, RD       paren

    # mint mark reference: en.wikipedia.org/wiki/Historical_United_States_mints
    # can have slash with two mints: e.g. #3985: 1938-D/S 5C Buffalo

    # Detail is just everything after the denom, or after date if no denom
    #   #5063: 1945-S 10C -->Micro S, FB<--
    #   #802036: 1814 10C -->STATESOF S.S. Central America #2 (with Pinch)<--
    # Search algorithm will handle this by calculating edit distance between the
    # multiple options and ranking them

    for entry in price_guide:
        description = entry['description']

        # get year
        year_search = YEAR.match(description)
        if year_search is None:
            year_short = None
            year_full = None
        else:
            # year_short is for search purposes, year_full is more descriptive
            year_short = year_search.group(1)       # e.g. 1825
            year_full = year_search.group(0)        # e.g. 1825/4/(2) for errors

        # get mint information
        mint_search = MINT.search(description)
        if mint_search is None:
            mint = None
        else:
            mint = mint_search.group(0)
            mint = mint.strip('-')

        # get denomination
        # get detail (everything after denomination, or year if no denom)
        denom_search = DENOM.search(description)
        if denom_search is None:
            denom = None
            # detail information would start at different points
            if year_short is None:
                detail = description    # no denomination no year
            else:
                year_span_end = year_search.regs[0][1]
                detail = description[year_span_end:]    # year but no denom
        else:
            denom = denom_search.group(0)
            denom_span_end = denom_search.regs[0][1]
            detail = description[denom_span_end:]

        # update entry with parsed description info
        entry['year_short'] = year_short
        entry['year_full'] = year_full
        entry['mint'] = mint
        entry['denom'] = denom
        entry['detail'] = detail

    return price_guide


def word_to_digit(input_string):
    """
    Help normalize the query string by replacing typed-out forms of numbers into
    digits. NOTE: this is only for numbers common to coins, and is not meant to
    be for all numbers

    :param input_string: string to change words to digits
    :return output_string: new string with digits in place of numbers
    """
    replacement_pairs = [
        ('one', '1'),
        ('two', '2'),
        ('three', '3'),
        ('four', '4'),
        (r'(?<!twenty )five', '5'),
        ('six', '6'),
        ('seven', '7'),
        ('eight', '8'),
        ('nine', '9'),
        ('ten', '10'),
        ('twenty', '20'),
        ('twenty five', '25'),
        ('fifty', '50')
    ]
    for pattern, replacement in replacement_pairs:
        output_string = re.sub(pattern, replacement, input_string)

    return output_string


def fold_denoms(query_str):
    """
    Given a user input string, normalize and 'fold' denominations into PCGS
    denomination category, less focus is given to colonial coins here, that's a
    to-do

    :param query_str: string queried by the user
    :return :
    """
    denominations_mapping = [
        # key is PCGS denomination
        (r'1/2C', r'(1 )?half( of (a|one) ?)? cents?'),  # half cent
        (r'1C', r'(1 cent|penny)'),  # 1 cent
        # (r'Cent', ['1C']),
        # (r' C ', ['1C'],
        (r'2C', r'2 cents? (silver)?'),  # 2 cent
        (r'3CS', r'3( ?C| cents?) silver'),  # flag this and ask if they mean silver or nickel
        (r'3CN', r'3 cents? nickel'),  # 3 cent nickel
        (r'5C', r'((?<!cent )nickel|5 cents?)'),  # 5 cent (nickel)
        (r'H10C', r'half dime'),  # half Dime
        (r'10C', '(dime|10 cents?)'),  # dime
        (r'20C', r'20 cents?'),  # 20 cent
        (r'25C', r'(quarter|25 cents?)'),  # quarter
        (r'50C', r'(?<!and a )(half dollar|50 cents?)'),  # half dollar
        (r'$1', r'(?<![02-9] )(?<!half )(1 )?dollar'),  # dollar coin
        (r'$2.50', r'(2 (and a half dollars?|dollars and 50 cents)|\$?2\.50)'),  # $2.50 gold
        (r'$3', r'3 dollars?'),  # $3 gold
        (r'$4', r'4 dollars?'),  # $4 gold
        (r'$5', r'5 dollars?'),  # $4 gold
        (r'$10', r'10 dollars?'),  # $10 gold
        (r'$20', r'20 dollars?'),  # $20 St. Gaudens Double Eagle
        (r'$25', r'25 dollars?'),  # Gold Eagle
        (r'$50', r'50 dollars?'),  # Gold Eagle
        # after all of the main ones are done
        (r'1C', r'(?<![0-9] )(?<!half )cent')     # lincoln cent, wheat cent, etc.
        # Colonials and special cases
        # (r'\'?Penny\'?', ['']),
        # (r'1P', ['']),
        # (r'1/2 ?P', ['']),
        # (r'3Pence', ['']),
        # (r'2Pence', ['']),
        # (r'6Pence', ['']),
        # (r'Shilling', ['']),
        # (r'Shilng', ['']),  # Special case, see PCGS#249
        # (r'1/24RL', ['']),  # see PCGS#49
        # (r'9 Den', ['']),  # French colonies, Deniers
        # (r'15 Den', ['']),
        # (r'30 Den', ['']),
        # (r'Farth', ['']),  # e.g. PCGS#256
        # (r'Sou', ['']),  # French Colonies
        # (r'Sol', ['']),  # French Colonies, see PCGS#167113
        # (r'1/2 Db', ['']),  # see PCGS#489
        # (r'1/2 (R', ['']),  # see PCGS#600506
        # (r'1/2 RL', ['']),
        # (r'Rial', ['']
    ]

    query_str = word_to_digit(query_str)

    for repl, pattern in denominations_mapping:
        query_str = re.sub(pattern, repl, query_str, flags=re.IGNORECASE)

    return query_str


def price_table(desig, prices_by_grade):
    """
    Generate printable table for price data

    :param desig: designation info, from coin_entry['desig'] for a given
        coin_entry in the price_guide ft
    :param prices_by_grade: coin_entry['prices'] for a given desig
    :return :
    """
    # get max length for every row to format table
    all_lens = []
    all_lens.extend([len(_) for _ in desig])
    for grade, prices in prices_by_grade.items():
        all_lens.append(len(str(grade)))
        for price in prices:
            if price is not None:
                all_lens.append(len(price))
            else:
                all_lens.append(4)      # len(str(None))
    max_len = max(all_lens)

    # format table parts
    header_row = f'%-{max_len+2}s%-{max_len+2}s%-{max_len+2}s\n' % \
                 ('grade', desig[0], desig[1])
    lines = '='*(max_len+1)
    midrule = lines + '|' + lines + '|' + lines + '\n'

    # table body
    rows = []
    for grade, (price0, price1) in prices_by_grade.items():
        row = f'%-{max_len + 2}s%-{max_len + 2}s%-{max_len + 2}s\n' % \
              (grade, price0, price1)
        rows.append(row)
    body = ''.join(rows)

    table = header_row + midrule + body

    return table


pcgs_scraper/pcgs_prices.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
pcgs_prices.py

Script to programmatically scrape all coin prices from www.pcgs.com/prices

Author: Ryan A. Mannion, 2020
github: ryanamannion
twitter: @ryanamannion
"""
import json
import time
import pickle
import argparse
from tqdm import tqdm
from datetime import datetime
from collections import defaultdict

import ft
from bs4 import BeautifulSoup

from pcgs_scraper.utils import request_page, non_ns_children

INDEX = 'https://www.pcgs.com'
PRICES = 'https://www.pcgs.com/prices'
BINS = ['grades-1-20', 'grades-25-60', 'grades-61-70']
GRADES = [1, 2, 3, 4, 6, 8, 10, 12, 15, 20, 25, 30, 35, 40, 45, 50, 53, 55, 58,
          60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70]


######################
# SCRAPING FUNCTIONS #
######################
def get_urls(page_url):
    """
    Gets urls for each subcategory on the /prices page, returns them as a
    dictionary where the keys are the categories, and the values are a tuple of
    (subcategory_name, subcategory_url)

    :return urls_by_category: (dict) dict of all urls on /prices page
    """
    # prep page html and soup
    page = request_page(page_url)
    soup = BeautifulSoup(page.text, 'html.parser')

    urls_by_category = defaultdict(list)

    # get all category urls from /prices page
    # first: get two columns which contain the boxes for each category
    columns = soup.find_all('div', class_='col-xs-12 col-sm-6')
    for column in columns:
        for box in non_ns_children(column, 'children'):
            for box_element in non_ns_children(box, 'children'):
                if 'class' in box_element.attrs and \
                        'coin-heading' in box_element.attrs['class']:
                    heading = box_element.text.strip()
                elif box_element.name == 'ul':
                    for list_item in non_ns_children(box_element, 'children'):
                        # presumably only has one tag as a child, <a>
                        category_href = list_item.contents[0].attrs['href']
                        category_url = INDEX + category_href
                        category_name = list_item.contents[0].text.strip()
                        this_category = (category_name, category_url)
                        urls_by_category[heading].append(this_category)
    # turn into a regular dict
    return dict(urls_by_category)


def get_prices(url, delay_s=1.5):
    """
    For a given URL, extract price information

    :param url: url for a page under www.pcgs.com/prices/detail/...
    :param delay_s: seconds to sleep to avoid response status 429, 1.5s worked
        in my testing, though I did not try any lower
    :return prices: a list of dictionaries representing each row in the table
    """

    # 1: prep for task
    time.sleep(delay_s)
    page = request_page(url)
    soup = BeautifulSoup(page.text, 'html.parser')

    # 2: determine which grades we are dealing with in the table
    for grade_bin in BINS:
        if grade_bin in url:
            grades = grade_bin

    # 3: find the table
    table = soup.find('table', class_='table-main')
    if table is None:
        # some pages like counterstamped colonials have no price information,
        # would crash script otherwise
        return []

    # 4: loop through table elements and scrape prices
    prices = []
    for element in non_ns_children(table, 'children'):
        if element.name == 'tbody':
            for child in non_ns_children(element, 'children'):
                # table has lots of elements, but the ones which contain coin
                # prices are colored bg-pale and bg-light, this makes IDing
                # them a lot easier
                if 'class' in child.attrs:      # check if it has class attr
                    if 'bg-pale' in child.attrs['class'] \
                            or 'bg-light' in child.attrs['class']:
                        # each row has 13 cells with two possible internal rows
                        cells = non_ns_children(child, 'children')
                        row_prices = []     # to collect all prices in this row
                        for i, cell in enumerate(cells):
                            if i == 0:          # cell 0 is the pcgs number
                                pcgs_num = cell.text.strip()
                                if len(pcgs_num) == 0:
                                    pcgs_num = None
                            elif i == 1:        # cell 1 is the description
                                description = cell.text.strip()
                            elif i == 2:        # cell 2 is the designation
                                desig = cell.text.split()
                            elif 3 <= i <= 12:  # rest are prices
                                this_cell_prices = cell.text.strip('▲▼').split()
                                # always make prices cell a double for ease
                                # KNOWN ISSUE: in some rare cases, only one
                                # price is shown and appears in the bottom row
                                if len(this_cell_prices) == 0:
                                    this_cell_prices = (None, None)
                                elif len(this_cell_prices) == 1:
                                    this_cell_prices.append(None)
                                this_cell_prices = tuple(this_cell_prices)
                                row_prices.append(this_cell_prices)
                        # after each cell is scraped, turn into a dict & append
                        row = {
                            'pcgs_num': pcgs_num,
                            'description': description,
                            'desig': desig,
                            'grades': grades,
                            'prices': row_prices,
                            'url': url
                        }
                        prices.append(row)
    return prices


def scrape_all():
    """
    Entire scraping process in one call:
        Step 1: Load www.pcgs.com/prices and get all URL information
        Step 2: Load each URL and scrape prices from its table
        Step 3: Save price information to pickle
    """

    # Step 1
    print(f"Getting URLs from {PRICES}...")
    urls_by_category = get_urls(PRICES)
    print("Success!")

    # Step 2
    print("Scraping price data by category...")
    prices = []
    for i, (category, subcategories) in enumerate(urls_by_category.items()):
        print(f"\tBeginning category {i + 1}/{len(urls_by_category.items())}:"
              f" {category}")
        for subcat, subcat_url in tqdm(subcategories):
            for grade_bin in BINS:
                # url defaults to most-active page first, but we want all the
                # grade information
                this_bin_url = subcat_url.replace('most-active', grade_bin)
                this_bin_url += '?pn=1&ps=-1'       # show all prices one page
                this_bin_prices = get_prices(this_bin_url, delay_s=1.0)
                prices.extend(this_bin_prices)
    print("Success!")

    # Step 3
    today = datetime.now()
    current_time = today.strftime("%d-%m-%Y-%H:%M:%S")
    filename = f'data/pcgs_prices_unprocessed-{current_time}.pkl'
    print(f"Saving price data to {filename}")
    pickle.dump(prices, open(filename, 'wb'))
    print(f"Success!")
    return filename


########################
# PROCESS SCRAPED DATA #
########################
def merge_grade_bins(filepath):
    """
    data from scrape_all is separated by grade_bins, combine into a single
    entry for each pcgs number

    The reason this is a separate function from the scrape function is to ensure
    that the scraped data is saved as soon as possible to avoid errors causing
    data loss after waiting for all the prices to be scraped

    :param filepath: path to scraped data pkl from scrape_all
    :return price_guide: (dict) lookup table for
    """
    # NOTE: there will be a lot of entries with a None pcgs_num, these are
    # typically the prices for full type sets of a certain coin on the price
    # detail page, which this script currently does not account for. To get
    # that information, capture the title of each subsection of the table and
    # you can relate it

    price_guide = {}

    scraped_data = pickle.load(open(filepath, 'rb'))
    by_pcgs_num = ft.indexBy('pcgs_num', scraped_data)

    for pcgs_num, entries in by_pcgs_num.items():
        if pcgs_num is None:
            continue        # see above comment

        # 1: merge price information into a single dict that points from # to $
        assert len(entries) == 3, 'Something went wrong, more than 3 bins'
        # make absolutely certain that prices are in the correct order, overkill
        temp_order = [None, None, None]
        desigs = []
        for entry in entries:
            desigs.append(entry['desig'])       # save desig for step 2
            if entry['grades'] == 'grades-1-20':
                temp_order[0] = entry['prices']
            elif entry['grades'] == 'grades-25-60':
                temp_order[1] = entry['prices']
            elif entry['grades'] == 'grades-61-70':
                temp_order[2] = entry['prices']
        this_num_prices = []        # prices for this pcgs number
        for grade_bin in temp_order:
            for price in grade_bin:
                this_num_prices.append(price)
        assert len(this_num_prices) == len(GRADES), \
            f'Wrong number of grades for PCGS#{pcgs_num}: ' \
            f'{len(this_num_prices)}'
        price_by_grade = dict(zip(GRADES, this_num_prices))

        # 2: Ensure that the desig is always two place if at least one is
        merged_desig = []       # start with len == 0
        for desig in desigs:
            if len(desig) > len(merged_desig):   # longest desig wins
                merged_desig = desig

        merged_entry = {
            'pcgs_num': pcgs_num,
            'desig': merged_desig,
            'prices': price_by_grade,
            'merged_from': entries,
        }
        price_guide[pcgs_num] = merged_entry

    print('Saving price guide to pkl and json files...')
    pickle.dump(price_guide, open('data/scraped_pcgs_prices.pkl', 'wb'))
    with open('data/scraped_pcgs_prices.json', 'w') as outfile:
        json.dump(price_guide, outfile)


def main():
    save_file = scrape_all()
    merge_grade_bins(save_file)


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument('--all', '-a', action='store_true',
                        help="scrape all coin prices, create lookup table")
    parser.add_argument('--scrape_only', '-s', action='store_true',
                        help="only scrape coin prices and save unprocessed "
                             "file")
    parser.add_argument('--process', '-p', action='store',
                        help="process only, specify path to .pkl file to "
                             "process and create lookup table from")

    args = parser.parse_args()

    if args.all is True:
        main()
    elif args.scrape_only is True:
        scrape_all()
    elif args.process is not None:
        merge_grade_bins(args.process)
    else:
        print('Please specify an option. Documentation available at '
              'https://github.com/ryanamannion/pcgs_prices')
